crates/store/re_types_core/src/loggable_batch.rs

```diff
@@ -39,64 +38,131 @@ pub trait ComponentBatch: LoggableBatch {
         Arrow2ListArray::<i32>::try_new(data_type, offsets.into(), array.to_boxed(), None)
             .map_err(|err| err.into())
     }
+
+    /// Returns the complete [`ComponentDescriptor`] for this [`ComponentBatch`].
+    ///
+    /// Every component batch is uniquely identified by its [`ComponentDescriptor`].
+    fn descriptor(&self) -> Cow<'_, ComponentDescriptor>;
+
+    /// The fully-qualified name of this component batch, e.g. `rerun.components.Position2D`.
+    ///
+    /// This is a trivial but useful helper for `self.descriptor().component_name`.
+    ///
+    /// The default implementation already does the right thing. Do not override unless you know
+    /// what you're doing.
+    /// `Self::name()` must exactly match the value returned by `self.descriptor().component_name`,
+    /// or undefined behavior ensues.
+    #[inline]
+    fn name(&self) -> ComponentName {
+        self.descriptor().component_name
+    }
+}
+
+/// Some [`ComponentBatch`], optionally with an overridden [`ComponentDescriptor`].
+///
+/// Used by implementers of [`crate::AsComponents`] to both efficiently expose their component data
+/// and assign the right tags given the surrounding context.
+pub struct MaybeOwnedComponentBatch<'a> {
```

Comment:
Maybe at this point this deserves a new name? I don't think it matters all that much, all those APIs will break in the most extraordinary ways once we move to eager serialization anyhow...

Reply:
That said, `DescribedComponentBatch` could be made to match what we use on the Python side, and maybe even the C++ side if we get lucky...

Reply:
agreed. Maybe simpler `ComponentBatchCowWithDesc`?